File: javascript/json-transform/src/functions/TransformerFunctionDigest.ts
import TransformerFunction from "./common/TransformerFunction";
import { ArgType } from "./common/ArgType";
import FunctionContext from "./common/FunctionContext";
import TextEncoding from "./common/TextEncoding";
import Base64 from "./utils/Base64";
import md5 from "./utils/md5";

const globalCrypto = typeof global.crypto !== "undefined" ? global.crypto : require("crypto");

function formatHex(a: ArrayBuffer) {
  return [...new Uint8Array(a)].map(x => x.toString(16).padStart(2, "0")).join("");
}

class TransformerFunctionDigest extends TransformerFunction {
  constructor() {
    super({
      argsSet: [
        { name: "algorithm", type: ArgType.String, defaultValue: "SHA-1" },
        { name: "format", type: ArgType.String, defaultValue: "BASE64" },
      ],
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    const str = await context.getString(null);
    if (str == null) {
      return null;
    }
    const algorithm = await context.getEnum("algorithm");
    if (!algorithm) return null;

    const digest =
      algorithm === "MD5"
        ? md5(str)
        : await globalCrypto.subtle.digest({ name: algorithm }, TextEncoding.encode(str, "ISO-8859-1"));
    switch (await context.getEnum("format")) {
      case "BASE64":
        return Base64.encode(new Uint8Array(digest), "basic");
      case "BASE64URL":
        return Base64.encode(new Uint8Array(digest), "url");
      default:
        return formatHex(digest);
    }
  }
}

export default TransformerFunctionDigest;

File: java/json-transform/src/main/java/ai/bizone/jsontransform/functions/TransformerFunctionDigest.java
package ai.bizone.jsontransform.functions;

import ai.bizone.jsontransform.functions.common.*;

import java.io.UnsupportedEncodingException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Base64;
import java.util.HexFormat;
import java.util.Map;

public class TransformerFunctionDigest extends TransformerFunction {
    public TransformerFunctionDigest() {
        super(FunctionDescription.of(
            Map.of(
            "algorithm", ArgumentType.of(ArgType.String).position(0).defaultValue("SHA-1"),
            "format", ArgumentType.of(ArgType.String).position(1).defaultValue("BASE64"),
            "charset", ArgumentType.of(ArgType.String).position(2).defaultValue("UTF-8")
            )
        ));
    }
    @Override
    public Object apply(FunctionContext context) {
        var str = context.getString(null);
        if (str == null) {
            return null;
        }
        var algorithm = context.getEnum("algorithm");
        var charset = context.getEnum("charset");
        try {
            var digest = MessageDigest.getInstance(algorithm).digest(str.getBytes(charset));
            return switch (context.getEnum("format")) {
                case "BASE64" -> Base64.getEncoder().encodeToString(digest);
                case "BASE64URL" -> Base64.getUrlEncoder().encodeToString(digest);
                default -> HexFormat.of().formatHex(digest);
            };
        } catch (NoSuchAlgorithmException | UnsupportedEncodingException e) {
            throw new RuntimeException(e);
        }
    }
}
